### PR TITLE
Add verify-spelling.sh for English words.

### LINKED ISSUE #545: Add verify-spelling.sh
### What would you like to be added?

https://github.com/kubernetes/website/blob/main/scripts/verify-spelling.sh

### Why is this needed?

NONE

### PR DESCRIPTION
**Pull request by yanggangtony**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add verify-spelling.sh to check english typos...

Ignore .golangci.yaml, 
```
./.golangci.yaml:22:6: "importas" is a misspelling of "imports"
./.golangci.yaml:50:2: "importas" is a misspelling of "imports"
```
because importas is the package name.
[golinters importas typos](https://github.com/golangci/golangci-lint/blob/cbc3a0c70ec14f15612caa056a805e229d64a529/pkg/golinters/importas.go#L7)

#### Which issue(s) this PR fixes:
Fixes #545

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: hack/verify-spelling.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+function check() {
+  #NOTE we use go install refer the go site :https://go.dev/doc/go-get-install-deprecation
+  go install github.com/client9/misspell/cmd/misspell@${TOOL_VERSION}
+  cd "${ROOT}"
+  # check spelling
+  echo "Checking spelling..."
+  ERROR_LOG="${TMP_DIR}/errors.log"
+
+  find . -type f | grep -v .golangci.yaml | grep -v vendor/ | xargs misspell >"${ERROR_LOG}"
```
Comment (wzshiming) [inline hack/verify-spelling.sh:56]: We should have a list of allowed words, not the whole document excluded.

Comment (yanggangtony) [reply hack/verify-spelling.sh:56]: thanks for review.

you means like 
```
skipping_file="${KUBE_ROOT}/scripts/.spelling_failures"
failing_packages=$(sed "s| | -e |g" "${skipping_file}")
```, 
And we add the ```.golangci.yaml``` and ```vendor/``` to it?

Actually before i did it like that,but error shows, return i update the code ..😂😂

Comment (wzshiming) [reply hack/verify-spelling.sh:56]: For more tools details see https://github.com/client9/misspell
Refer #514 to add `verify-spelling.sh` and `update-spelling.sh`

File: hack/verify-all.sh [hunk 2]
```diff
@@ -66,6 +66,11 @@ if [[ "${VERIFY_SHELL_FORMAT:-true}" == "true" ]]; then
   "${ROOT_DIR}"/hack/verify-shell-format.sh || failed+=(shell-format)
 fi
 
+if [[ "${VERIFY_SHELL_SPELLING:-true}" == "true" ]]; then
+  echo "[*] Verifying shell spelling..."
```
Comment (wzshiming) [inline hack/verify-all.sh:70]:
```suggestion
  echo "[*] Verifying spelling..."
```

File: hack/verify-spelling.sh (hunk 1 above)
Comment (yanggangtony) [reply hack/verify-spelling.sh:56]: thanks .
I will do it later.

File: hack/update-spelling.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+function check() {
+  #NOTE we use go install refer the go site :https://go.dev/doc/go-get-install-deprecation
+  go install github.com/client9/misspell/cmd/misspell@${TOOL_VERSION}
+  cd "${ROOT}"
+  # check spelling
+  echo "Checking spelling..."
+  ERROR_LOG="${TMP_DIR}/errors.log"
+
+  find . -type f | grep -v vendor/ | xargs misspell -i $(sed <"${allowd_spelling_words}" "s| |,|g") . >"${ERROR_LOG}"
```
Comment (wzshiming) [inline hack/update-spelling.sh:45]:
```suggestion
  ignore="$(tr  <"${allowd_spelling_words}" '\n' ',')"
  find . -type f | grep -v vendor/ | xargs misspell -i "${ignore}" .
```

File: hack/allowd_spelling_words [hunk 4]
```diff
@@ -0,0 +1 @@
+importas,
```
Comment (wzshiming) [inline hack/allowd_spelling_words:1]:
```suggestion
importas

```

File: hack/update-spelling.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+# cleanup
+exitHandler() (
+  echo "Cleaning up..."
+  rm -rf "${TMP_DIR}"
+)
+trap exitHandler EXIT
+
+function check() {
+  #NOTE we use go install refer the go site :https://go.dev/doc/go-get-install-deprecation
+  go install github.com/client9/misspell/cmd/misspell@${TOOL_VERSION}
```
Comment (wzshiming) [inline hack/update-spelling.sh:39]: Refer to https://github.com/kubernetes-sigs/kwok/blob/77084dafb31221339996078e319501e4668c7d74/hack/update-shell-format.sh#L32C1-L41

**Comment by yanggangtony** [comment]
So fast for review...😀😀

**Comment by yanggangtony** [comment]
vendor dir is not exclude..
i will test it againt.
wait a moment for rereviw.

File: hack/verify-spelling.sh [hunk 6]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
+
```
Comment (wzshiming) [inline hack/verify-spelling.sh:2]:
```suggestion
```

File: hack/update-spelling.sh [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+TOOL_VERSION="v0.3.4"
+
+# cd to the root path
+ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd -P)"
+allowd_spelling_words="${ROOT}/hack/allowd_spelling_words"
```
Comment (wzshiming) [inline hack/update-spelling.sh:24]:
```suggestion
allowd_spelling_words="${ROOT}/hack/allowd_spelling_words.txt"
```

File: hack/update-spelling.sh [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+elif command -v "${GOPATH}/bin/misspell"; then
+  COMMAND=("${GOPATH}/bin/misspell")
+else
+  #NOTE we use go install refer the go site :https://go.dev/doc/go-get-install-deprecation
+  go install github.com/client9/misspell/cmd/misspell@${TOOL_VERSION}
+  COMMAND=(misspell)
+fi
+
+function check() {
+  cd "${ROOT}"
```
Comment (wzshiming) [inline hack/update-spelling.sh:48]:
```suggestion
```

File: hack/update-spelling.sh [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  # check spelling
+  echo "Checking spelling..."
+  ERROR_LOG="${TMP_DIR}/errors.log"
```
Comment (wzshiming) [inline hack/update-spelling.sh:51]:
```suggestion
```

File: hack/update-spelling.sh [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+  ignore="$(tr <"${allowd_spelling_words}" '\n' ',')"
+  "${COMMAND[@]}" -i "${ignore}" "$(find . -not \( -path ./vendor\* -o -path ./.git\* \))"
+
+  if [[ -s "${ERROR_LOG}" ]]; then
+    echo "Found spelling errors!" >&2
+    cat "${ERROR_LOG}"
+  fi
```
Comment (wzshiming) [inline hack/update-spelling.sh:59]:
```suggestion
```

File: hack/update-spelling.sh [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+check || exit 1
```
Comment (wzshiming) [inline hack/update-spelling.sh:62]:
```suggestion
cd "${ROOT_DIR}" && check || exit 1
```

File: hack/verify-all.sh [hunk 12]
```diff
@@ -66,6 +66,11 @@ if [[ "${VERIFY_SHELL_FORMAT:-true}" == "true" ]]; then
   "${ROOT_DIR}"/hack/verify-shell-format.sh || failed+=(shell-format)
 fi
 
+if [[ "${VERIFY_SHELL_SPELLING:-true}" == "true" ]]; then
+  echo "[*] Verifying spelling..."
+  "${ROOT_DIR}"/hack/verify-spelling.sh || failed+=(shell-spelling)
+fi
```
Comment (wzshiming) [inline hack/verify-all.sh:72]:
```suggestion
if [[ "${VERIFY_SPELLING:-true}" == "true" ]]; then
  echo "[*] Verifying spelling..."
  "${ROOT_DIR}"/hack/verify-spelling.sh || failed+=(spelling)
fi
```

File: hack/update-all.sh [hunk 13]
```diff
@@ -51,6 +51,11 @@ if [[ "${UPDATE_SHELL_FORMAT:-true}" == "true" ]]; then
   "${ROOT_DIR}"/hack/update-shell-format.sh || failed+=(shell-format)
 fi
 
+if [[ "${UPDATE_SHELL_SPELLING:-true}" == "true" ]]; then
+  echo "[*] Update spelling..."
+  "${ROOT_DIR}"/hack/update-spelling.sh || failed+=(shell-spelling)
+fi
```
Comment (wzshiming) [inline hack/update-all.sh:57]:
```suggestion
if [[ "${UPDATE_SPELLING:-true}" == "true" ]]; then
  echo "[*] Update spelling..."
  "${ROOT_DIR}"/hack/update-spelling.sh || failed+=(spelling)
fi
```

File: hack/verify-spelling.sh [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+# Copyright 2019 The Kubernetes Authors.
```
Comment (wzshiming) [inline hack/verify-spelling.sh:3]:
```suggestion
# Copyright 2023 The Kubernetes Authors.
```

File: hack/update-spelling.sh [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+  echo "Cleaning up..."
+  rm -rf "${TMP_DIR}"
+)
+trap exitHandler EXIT
+
+COMMAND=()
+if command -v misspell; then
+  COMMAND=(misspell)
+elif command -v "${GOPATH}/bin/misspell"; then
+  COMMAND=("${GOPATH}/bin/misspell")
```
Comment (wzshiming) [inline hack/update-spelling.sh:40]:
```suggestion
elif command -v "${ROOT_DIR}/bin/misspell"; then
  COMMAND=("${ROOT_DIR}/bin/misspell")
```

File: hack/update-spelling.sh [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+else
+  #NOTE we use go install refer the go site :https://go.dev/doc/go-get-install-deprecation
+  go install github.com/client9/misspell/cmd/misspell@${TOOL_VERSION}
+  COMMAND=(misspell)
```
Comment (wzshiming) [inline hack/update-spelling.sh:44]:
```suggestion
  GOBIN="${ROOT_DIR}/bin/" go install github.com/client9/misspell/cmd/misspell@${TOOL_VERSION}
  COMMAND=("${ROOT_DIR}/bin/misspell")
```

**Review by wzshiming — CHANGES_REQUESTED** [review]

**Comment by yanggangtony** [comment]
Good suggestion.
Update later.

File: hack/update-spelling.sh [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+  #NOTE we use go install refer the go site :https://go.dev/doc/go-get-install-deprecation
+  GOBIN="${ROOT_DIR}/bin/" go install github.com/client9/misspell/cmd/misspell@${TOOL_VERSION}
+  COMMAND=("${ROOT_DIR}/bin/misspell")
+fi
+
+function check() {
+  # check spelling
+  echo "Checking spelling..."
+  ignore="$(tr <"${allowd_spelling_words}" '\n' ',')"
+  "${COMMAND[@]}" -i "${ignore}" "$(find . -not \( -path ./vendor\* -o -path ./.git\* \))"
```
Comment (wzshiming) [inline hack/update-spelling.sh:40]:
```suggestion
  ignore="$(tr <"${allowd_spelling_words}" '\n' ',')"
  "${COMMAND[@]}" -w -i "${ignore}" "$(find . -not \( -path ./vendor\* -o -path ./.git\* \))"
```

File: hack/update-spelling.sh [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+ROOT_DIR="$(realpath "$(dirname "${BASH_SOURCE[0]}")"/..)"
+allowd_spelling_words="${ROOT_DIR}/hack/allowd_spelling_words"
+
+COMMAND=()
+if command -v misspell; then
+  COMMAND=(misspell)
+elif command -v "${ROOT_DIR}/bin/misspell"; then
+  COMMAND=("${ROOT_DIR}/bin/misspell")
+else
+  #NOTE we use go install refer the go site :https://go.dev/doc/go-get-install-deprecation
```
Comment (wzshiming) [inline hack/update-spelling.sh:31]:
```suggestion
```

File: hack/update-spelling.sh [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+TOOL_VERSION="v0.3.4"
+
+ROOT_DIR="$(realpath "$(dirname "${BASH_SOURCE[0]}")"/..)"
+allowd_spelling_words="${ROOT_DIR}/hack/allowd_spelling_words"
```
Comment (wzshiming) [inline hack/update-spelling.sh:23]:
```suggestion
allowed_spelling_words="${ROOT_DIR}/hack/spelling.txt"
```

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: hack/update-spelling.sh [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+fi
+
+function check() {
+  # check spelling
+  echo "Checking spelling..."
+  ignore="$(tr <"${allowd_spelling_words}" '\n' ',')"
+  "${COMMAND[@]}" -i "${ignore}" "$(find . -not \( -path ./vendor\* -o -path ./.git\* \))"
+}
+
+cd "${ROOT_DIR}" && check || exit 1
```
Comment (wzshiming) [inline hack/update-spelling.sh:43]:
```suggestion
cd "${ROOT_DIR}"
 || exit 1
```

File: hack/update-spelling.sh (hunk 17 above)
Comment (yanggangtony) [reply hack/update-spelling.sh:40]: Good suggestion.
with -w flag , So misspell can override the file.
And then ```git --no-pager diff --exit-code``` can check the diff of file.

Update later.

File: hack/update-spelling.sh [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+elif command -v "${ROOT_DIR}/bin/misspell"; then
+  COMMAND=("${ROOT_DIR}/bin/misspell")
+else
+  GOBIN="${ROOT_DIR}/bin/" go install github.com/client9/misspell/cmd/misspell@${TOOL_VERSION}
+  COMMAND=("${ROOT_DIR}/bin/misspell")
+fi
+
+function update() {
+  ignore="$(tr <"${allowed_spelling_words}" '\n' ',')"
+  "${COMMAND[@]}" -w -i "${ignore}" "$(find . -not \( -path ./vendor\* -o -path ./.git\* \))"
```
Comment (wzshiming) [inline hack/update-spelling.sh:37]:
```suggestion
  local ignore="$(tr <"${allowed_spelling_words}" '\n' ',')"
  mapfile -t files < <(find . \( \
    -iname "*.md" -o \
    -iname "*.sh" -o \
    -iname "*.go" -o \
    -iname "*.tpl" -o \
    -iname "*.yaml" -o \
    -iname "*.yml" \
    \) \
    -not \( \
    -path ./.git/\* -o \
    -path ./vendor/\* -o \
    -path ./demo/node_modules/\* -o \
    -path ./site/themes/\* \
    \))
  "${COMMAND[@]}" -locale US -w -i "${ignore}" "${files[@]}"
```

File: pkg/log/logger_ctl.go [hunk 22]
```diff
@@ -139,19 +139,19 @@ func (c *ctlHandler) Handle(r slog.Record) error {
...
-	return colour{
+func newColour(c ctc.Color, msg string) color {
+	return color{
 		renderer: fmt.Sprintf("%s%s%s", c, msg, ctc.Reset),
 		width:    stringWidth(msg),
 	}
 }
 
-var levelColour = map[string]colour{
+var levelColour = map[string]color{
```
Comment (wzshiming) [inline pkg/log/logger_ctl.go:154]:
```suggestion
var levelColor = map[string]color{
```

Comment (yanggangtony) [reply pkg/log/logger_ctl.go:154]: good, why the misspell not found that....😂😂

**Review by wzshiming — APPROVED** [review]
/approve
/lgtm
/label tide/merge-method-squash

**Pull request merged**